### PR TITLE
Opimtize key tree stemming by using maps instead of sets

### PR DESCRIPTION
**Pull request by nickva**
sets to due to a compiler bug in 22 consumes too much memory and is slower on Erlang 22+

Reproducer: https://gist.github.com/nickva/ddc499e6e05678faf20d344c6e11daaa

With sets:
```
couch_key_tree:gen_and_stem().
{8,6848.812683105469}
```

With maps:
```
couch_key_tree:gen_and_stem().
{0,544.000732421875}
```

### CODE REVIEW

**Comment by davisp** [comment]
I'd also note that this improves performance across the board. Using maps on 20 and 21 also improves against current baseline.

+1

**Comment by nickva** [comment]
Trying the optimization with a production revision tree with 6638 live conflicts.

```
length(couch_key_tree:get_all_leafs(Tree)).
6638
```

Measured time to run `couch_key_tree:stem(Tree, 1000)`

### Before this commit


| Erlang Version  |  Time (sec) | Memory (GBs) |
|----------------|-------|---------|
| 20 | 221 | 1.2 |
| 23 | 1382 | 45 |

### After this commit

| Erlang Version  |  Time (sec) | Memory (GBs) |
|----------------|-------|---------|
| 20 | 3 | 1.2 |
| 23 | 3 | 1.3 |

In summary, on Erlang 23, we got a nice **460x** speedup and **34x** memory usage reduction

**Comment by kocolosk** [comment]
Now you're just showing off 😉 Nice work!
